### PR TITLE
8273485: Deadlock when also using Swing and exiting Fullscreen on Mac

### PR DESCRIPTION
**Pull request by FlorianKirmaier**
When using Swing it's possible to generate a Deadlock.
 It's related to the nested eventloop started in enterFullScreenExitingLoop - and the RenderLock aquired when using setView in Scene.
 Sample Programm and Threaddump are added to the ticket.

Removing the nested loop fixes the Problem. 
I hope this doesn't have any side effect - so far i don't know of any.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8273485](https://bugs.openjdk.org/browse/JDK-8273485): Deadlock when also using Swing and exiting Fullscreen on Mac (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/622/head:pull/622` \
`$ git checkout pull/622`

Update a local copy of the PR: \
`$ git checkout pull/622` \
`$ git pull https://git.openjdk.org/jfx.git pull/622/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 622`

View PR using the GUI difftool: \
`$ git pr show -t 622`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/622.diff">https://git.openjdk.org/jfx/pull/622.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/622#issuecomment-915124700)

### CODE REVIEW

**Review by kevinrushforth — CHANGES_REQUESTED** [review]
I am almost certain that this proposed fix will cause problems. At the very least we will now have mismatched enter / exit nested loop calls. The nested loop for full screen on Mac was done for a reason, so you will need to track down what that reason is and either verify that it is no longer needed or else solve that some other way.

Also, you don't explain why removing the nested event loop is the right fix for the deadlock.

Finally, once you address the above two points, you will need to turn the test program attached to the JBS bug into an automated test.

**Comment by FlorianKirmaier** [comment]
Thank you for the feedback - I will start writing an unit test for it because it will help on all further development regarding this issue.

**Comment by FlorianKirmaier** [comment]
I've now added a unit test.

I've adapted the solution to entirely remove the nested event loop, for the fullscreen. So far everything seems to work.
I've tested it with a simple application to check how the fullscreen behaves.
The logic seems to be added before the beginning of the VCS / 2013
Is there away to look into the history before 2013? Maybe there is a hint, why this was added.
Otherwise i would guess that it is not too important, and it is just an unnecessary overcomplicated solution.

**Review by kevinrushforth — CHANGES_REQUESTED** [review]
> I've adapted the solution to entirely remove the nested event loop, for the fullscreen. So far everything seems to work.
> I've tested it with a simple application to check how the fullscreen behaves.
> The logic seems to be added before the beginning of the VCS / 2013
> Is there away to look into the history before 2013? Maybe there is a hint, why this was added.

I tracked down the bug fix that added the nested event loop (looking at the history of the closed-source changes, since this fix was done prior to open-sourcing glass). It was added to fix [JDK-8126842](https://bugs.openjdk.java.net/browse/JDK-8126842).

As I suspected, the nested even loop is still necessary. If you run the test program attached to JDK-8126842 with your patch applied, make the window full-screen using the green button, then click the close button, the program will terminate with an exception.

**Comment by kevinrushforth** [comment]
You can also see the exception with _any_ JavaFX program, enter full-screen with the green button, move the mouse to the upper-left corner to reveal the window decorations, and press the red button to exit the program.

I think you need to look for a different solution to your deadlock, one that preserves the existing nested event loop.

**Comment by FlorianKirmaier** [comment]
Thank you for searching the original reason for the nested event loop!

I've now tested the class from the ticket, and for mit it works. The only strange thing is, that i get a "beep" sound when closing. Which isn't optimal but not the reported crash. Which is kinda unfortunate, because otherwise it might be easy to write a unit test.
Did you test the latest version of the PR?

I guess now i have to think of another fix, or on how do adapt this one.

**Comment by FlorianKirmaier** [comment]
@kevinrushforth 
Did you test really test my change and was able to reproduce the JDK-8126842 with it?
On my system everything works - except the small "beep" sound.

After investigating it, I've made another small change - which fixes the beep sound. It also makes it closer to the implementation for iOS. I've changed the `waitUntilDone` value to true, when closing the window.

I think removing the nestedEventLoops should be preferred over another fix - in my experience the nested event loops are the cause of so many rare and very complicated bugs.

**Comment by kevinrushforth** [comment]
I spoke too soon. I was running a few other manual tests, and your latest fix completely breaks the normal exit from full screen mode. To reproduce this:

1. Run `java hello.HelloFullscreenToggle`
2. Press the `Toggle Fullscreen` button to enter full-screen mode
3. Try to exit full screen mode, either by pressing the `ESC` key or by pressing the `Toggle Fullscreen` button again.
BUG: you can't exit full-screen mode.

**Review by kevinrushforth — COMMENTED** [review]
What version of macOS did you test on? I was running on 10.15.7 (Catalina). I just tried it on macOS 11.5 (Big Sur) and the exception does not happen, and on 10.14.6 (Mojave) where is does. To summarize, when calling `Platform.exit` while in full-screen mode:

macOS 10.14.6 : exception
macOS 10.15.7 : exception
macOS 11.5 : OK

File: modules/javafx.graphics/src/main/native-glass/mac/GlassViewDelegate.m [hunk 1]
```diff
@@ -1344,9 +1342,7 @@ - (void)exitFullscreenWithAnimate:(BOOL)animate
     {
         if (self->nativeFullScreenModeWindow)
         {
-            [self->nativeFullScreenModeWindow performSelector:@selector(toggleFullScreen:) withObject:nil];
-            // wait until the operation is complete
-            [GlassApplication enterFullScreenExitingLoop];
+            [[self->nsView window] toggleFullScreen:self];
```
[the remark this answers is not in the document]
Comment (FlorianKirmaier) [reply modules/javafx.graphics/src/main/native-glass/mac/GlassViewDelegate.m:1345]: With the previous version, I get a "beep" sound when closing a window.
But with the changed line, it works for me and I don't get any issues.
The beep sound also was the original reason, why I've removed the line.

**Comment by FlorianKirmaier** [comment]
I've been using macOS 11.6. It is also not so easy for me to test another version.

I guess the exception can easily be "fixed" by checking whether the application is null, in the `Screen.initScreens` method. Should I change it this way?

**Comment by FlorianKirmaier** [comment]
I've now added a check for null when calling the initScreens method. This should fix the null pointer exception.

File: modules/javafx.graphics/src/main/native-glass/mac/GlassViewDelegate.m [hunk 2]
```diff
@@ -1344,9 +1342,6 @@ - (void)exitFullscreenWithAnimate:(BOOL)animate
     {
         if (self->nativeFullScreenModeWindow)
         {
-            [self->nativeFullScreenModeWindow performSelector:@selector(toggleFullScreen:) withObject:nil];
-            // wait until the operation is complete
-            [GlassApplication enterFullScreenExitingLoop];
```
[the remark this answers is not in the document]
Comment (FlorianKirmaier) [reply modules/javafx.graphics/src/main/native-glass/mac/GlassViewDelegate.m:1349]: This was further discussed in the discussion below. So this is no longer up to date and therefore resolved.

**Comment by christianheilmann** [comment]
How to proceed to get this PR reviewed?

**Comment by FlorianKirmaier** [comment]
I've just merged it with the latest version in master.

**Comment by FlorianKirmaier** [comment]
I would like to keep this PR. 
I've worked weeks on it, to make this work.
It's now been tested quite a long in production.

I understand that this is hard to test.
But It removes code, adds a unit test, fixes a reproducible bug, and has been used for a long time in a Real World application. 
That's usually the kind of PR we want to accept.

**Comment by FlorianKirmaier** [comment]
Annually reminder that this pull request still exists.

Removing an unneeded NestedEventLoop is a huge reduction in complexity!

**Comment by FlorianKirmaier** [comment]
I guess there is no way to reopen pull requests, except creating new ones?

**Comment by beldenfox** [comment]
I took a look at this since I'm trying to better understand the Mac fullscreen code (there's a bunch of cruft I would like to remove).

The stack trace in the original bug report confused me. In `WindowStage.setScene` the window be should taken out of fullscreen mode before setScene acquires the renderLock and sets the view to null. Exiting fullscreen is failing because at this point in the close process the stage is hidden and you can't change the fullscreen state of a stage while it's hidden (see `WindowStage.applyFullScreen`).

I think it's a bug in the core Java code that you can hide a fullscreen window without having it automatically exit fullscreen. Hiding a fullscreen window doesn't even work on the Mac since the OS just ignores the `orderOut:` call.

As an experiment I tweaked the Mac code to exit fullscreen when the window is hidden using exactly the same code as in _setView and it avoided the deadlock state reported in the original bug. I haven't done a full test run with the changes.

**Comment by andy-goryachev-oracle** [comment]
> As an experiment I tweaked the Mac code to exit fullscreen when the window is hidden using exactly the same code as in _setView and it avoided the deadlock state reported in the original bug.

+1

**Comment by FlorianKirmaier** [comment]
It's nice to see there are other ways to fix the problem.

The approach of this PR had the big advantage of highly reducing the complexity.
Nested Event Loops are highly cursed - and removing one of them probably solves complicated issues at once.
We had so many issues with nested event loops - and in my opinion - they should even be removed entirely in the long run.

**Comment by kevinrushforth** [comment]
> I guess there is no way to reopen pull requests, except creating new ones?

There is, using the Skara `/open` command. In this case, I think the approach suggested by @beldenfox might be preferable.

**Comment by beldenfox** [comment]
> The approach of this PR had the big advantage of highly reducing the complexity.

Nested event loops are complicated and I agree we should ditch them when we can. But in this case Apple saddled us and themselves with complexity by making fullscreen transitions asynchronous. I did a quick Google search and had no problem finding problems. The Electron folks discovered that if you call toggleFullscreen during the fullscreen transition you get "weird behavior (incl. phantom windows)" and I was able to reproduce this. Who knows what other strangeness can happen. Anything we do to make toggleFullscreen behave more synchronously reduces our exposure to Apple bugs (and synchronous behavior is what JavaFX expects) so I think we'll just have to grit our teeth and deal with the nested event loops.

**Pull request closed without merging**